In [25]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.preprocessing import LabelEncoder

# **Part A: Dataset Understanding**

### Q1.Load the dataset and display the first five records.

In [2]:
df=pd.read_csv('Dataset2.csv')
df.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


### Q2.Determine the number of rows and columns in the dataset.



In [3]:
print('(Rows,Columns):',df.shape)

(Rows,Columns): (750, 10)


### Q3.Display all column names.


In [4]:
print('\nColumns:')
print(df.columns)


Columns:
Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


### Q4.Identify numerical and categorical features.

In [5]:
numerical_features = df.select_dtypes(include=np.number).columns.tolist()
categorical_features = df.select_dtypes(include='object').columns.tolist()

print('Numerical Features:', numerical_features)
print('Categorical Features:', categorical_features)

Numerical Features: ['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks', 'MonthlySpend']
Categorical Features: ['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed']


### Q5.Check whether the dataset contains missing values.

In [6]:
print('Missing values per column:')
print(df.isnull().sum())

Missing values per column:
UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64


# **Part B: Exploratory Data Analysis**

### Q6.Calculate the average age of users.

In [7]:
print('Average Age of the Users:', round(df['Age'].mean(),2))

Average Age of the Users: 41.82


### Q7.Determine the average watch hours per week.


In [8]:
print('Average watch hours per week:', round(df['WatchHoursPerWeek'].mean(),2))

Average watch hours per week: 14.24


### Q8.Find the average monthly spending of users.

In [9]:
print('Average monthly spending of users:', round(df['MonthlySpend'].mean(),2))

Average monthly spending of users: 689.91


### Q9.Count the number of users in each subscription category.


In [10]:
print('Number of users in each SubscriptionType:',df['SubscriptionType'].value_counts())

Number of users in each SubscriptionType: SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


### Q10.Determine the percentage of users who renewed their subscriptions.


In [11]:
renew = df['SubscriptionRenewed'].value_counts()
total_users = len(df)
percentage = (renew.get('Yes', 0) / total_users) * 100
print(f"Percentage of users who renewed their subscriptions: {percentage:.2f}%")

Percentage of users who renewed their subscriptions: 46.27%


# **Part C: Data Preparation**

### Q11.Convert categorical features into numerical form.


In [12]:
le = LabelEncoder()
for feature in categorical_features:
    df[feature] = le.fit_transform(df[feature])
print("DataFrame after converting categorical features:",display(df.head()))

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,0,0,23,5,1,13,353,0
1,1002,55,1,0,9,4,2,14,317,1
2,1003,49,1,0,8,3,1,16,309,0
3,1004,39,0,1,19,5,2,45,833,1
4,1005,38,0,1,23,5,5,24,804,1


DataFrame after converting categorical features: None


### Q12.Define the feature set (X) and target variable (y) for subscription renewal prediction.


In [13]:
X = df.drop(['UserID', 'SubscriptionRenewed'], axis=1)
y = df['SubscriptionRenewed']

print('Features (X) head:')
display(X.head())
print('\nTarget (y) head:')
display(y.head())

Features (X) head:


,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend
0,22,0,0,23,5,1,13,353
1,55,1,0,9,4,2,14,317
2,49,1,0,8,3,1,16,309
3,39,0,1,19,5,2,45,833
4,38,0,1,23,5,5,24,804



Target (y) head:


,SubscriptionRenewed
0,0
1,1
2,0
3,1
4,1


### Q13.Split the dataset into training and testing sets.


In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

X_train shape: (600, 8)
X_test shape: (150, 8)
y_train shape: (600,)
y_test shape: (150,)


# **Part D: Decision Tree Classification**

### Q14.Train a Decision Tree model to predict whether a user will renew their subscription.


In [19]:
dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train,y_train)
print("Decision Tree model trained successfully.")

Decision Tree model trained successfully.


### Q15.Evaluate the model using accuracy.


In [20]:
pred = dt.predict(X_test)
print('Decision Tree Accuracy:', accuracy_score(y_test,pred))

Decision Tree Accuracy: 0.56


### Q16.Generate and interpret the confusion matrix.

In [21]:
print('\nConfusion Matrix')
print(confusion_matrix(y_test,pred))

print("\nInterpretation:")
print(f"True Negatives (Correctly predicted 'No renewal'): {cm[0, 0]}")
print(f"False Positives (Incorrectly predicted 'Yes renewal' when it was 'No'): {cm[0, 1]}")
print(f"False Negatives (Incorrectly predicted 'No renewal' when it was 'Yes'): {cm[1, 0]}")
print(f"True Positives (Correctly predicted 'Yes renewal'): {cm[1, 1]}")


Confusion Matrix
[[48 34]
 [32 36]]

Interpretation:
True Negatives (Correctly predicted 'No renewal'): 48
False Positives (Incorrectly predicted 'Yes renewal' when it was 'No'): 34
False Negatives (Incorrectly predicted 'No renewal' when it was 'Yes'): 32
True Positives (Correctly predicted 'Yes renewal'): 36


# **Part E: K-Nearest Neighbors (KNN)**

### Q17.Train a KNN classifier with K = 5.


In [23]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train,y_train)
print("KNN trained successfully.")

KNN trained successfully.


### Q18.Compare the accuracy of KNN with the Decision Tree model.


In [24]:
knn_pred = knn.predict(X_test)
knn_accuracy = accuracy_score(y_test, knn_pred)

dt_pred = dt.predict(X_test)
dt_accuracy = accuracy_score(y_test, dt_pred)

if knn_accuracy > dt_accuracy:
    print("\nKNN model has higher accuracy.")
elif dt_accuracy > knn_accuracy:
    print("\nDecision Tree model has higher accuracy.")
else:
    print("\nBoth KNN and Decision Tree models have the same accuracy.")


KNN model has higher accuracy.


# **Part F: Linear Regression**

### 19.Train a Linear Regression model to predict monthly spending.


In [26]:
X_lr = df.drop(['UserID', 'SubscriptionRenewed', 'MonthlySpend'], axis=1)
y_lr = df['MonthlySpend']
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_lr, y_lr, test_size=0.2, random_state=42)
lr = LinearRegression()
lr.fit(X_train_reg, y_train_reg)
print("Linear Regression model trained successfully.")

Linear Regression model trained successfully.


### 20.Predict the monthly spending for a new user and interpret the result.



In [29]:
amount = lr.predict([[30,0,0,15,3,0,10]])
print('Predicted monthly spending for the new user:', amount[0])

Predicted monthly spending for the new user: 346.78228754457257


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


# **Business Reflection Questions**

### 1. Which factors appear to influence subscription renewal the most?  


In [30]:
feature_importances = dt.feature_importances_
feature_names = X.columns

importance_df = pd.DataFrame({
    'Feature': feature_names,
    'Importance': feature_importances
})

importance_df = importance_df.sort_values(by='Importance', ascending=False)

print("Factors influencing subscription renewal (from Decision Tree model):")
display(importance_df)

Factors influencing subscription renewal (from Decision Tree model):


,Feature,Importance
7,MonthlySpend,0.218895
6,AdClicks,0.199287
0,Age,0.192702
3,WatchHoursPerWeek,0.170035
5,FavoriteGenre,0.085444
4,DevicesUsed,0.070835
1,Gender,0.047127
2,SubscriptionType,0.015675


### 2. Why is subscription renewal a classification problem?  
Subscription renewal is a classification problem because the goal is to predict a categorical outcome: whether a user will renew their subscription ('Yes') or not ('No'). There are no continuous values to predict; instead, we are classifying users into one of two distinct groups. This binary nature (renew/not renew) is the hallmark of a binary classification problem.

### 3. Why is monthly spending a regression problem?  
Monthly spending is considered a regression problem because the target variable, 'MonthlySpend', is a continuous numerical value. Unlike classification, where the goal is to predict a discrete category, regression aims to predict a quantity that can take on any value within a given range. In this case, we are trying to predict the actual dollar amount a user spends per month, which is a continuous scale.

### 4. Which algorithm performed better for renewal prediction?  

Comparing the accuracy scores:
- Decision Tree Accuracy: 0.56
- KNN Accuracy: 0.63

The K-Nearest Neighbors (KNN) algorithm with K=5 performed better for renewal prediction, achieving an accuracy of 0.63 compared to the Decision Tree's accuracy of 0.56.

### 5. How could the platform use these predictions to improve customer retention?  

The platform can leverage these predictions to improve customer retention in several ways:

1.  **Proactive Intervention for Renewal Prediction:**
    *   By identifying users at high risk of not renewing their subscription, the platform can implement targeted retention strategies. This could include sending personalized offers, discounts, exclusive content access, or even direct outreach from customer support to address any issues or gather feedback.

2.  **Personalized Marketing Based on Monthly Spending:**
    *   The linear regression model predicting monthly spending can help segment users based on their spending potential. For users predicted to have high monthly spending, the platform can offer premium features, upsell opportunities, or exclusive bundles. For those with lower predicted spending, promotional offers or tiered subscription options might be more effective.

3.  **Feature Optimization based on Influencing Factors:**
    *   The feature importances from the Decision Tree model indicate which factors most influence subscription renewal. The platform can focus on improving or optimizing these areas. For example, if 'AdClicks' is a significant factor, the platform might analyze ad content and placement to improve user experience, or if 'WatchHoursPerWeek' is important, they might focus on content recommendations to increase engagement.

4.  **Targeted Content and Device Recommendations:**
    *   If 'FavoriteGenre' or 'DevicesUsed' are significant, the platform can tailor content recommendations or optimize the user interface for preferred devices, leading to a more satisfying experience and higher retention.

5.  **Understanding User Behavior:**
    *   Analyzing the characteristics of users who are predicted to churn can help the platform understand common pain points or dissatisfaction signals. This insight can then inform broader product development and service improvements to benefit all users.